src/fermilib/utils/_graph.py

```diff
@@ -0,0 +1,198 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from __future__ import absolute_import
+
+import queue
+
+
+class Node:
+    """Graph node
+
+    These graph nodes may be used to store data or other attributes within
+    the Graph structure.
+    """
+    def __init__(self, value=None):
+        """Build a graph node initialized with generic value"""
+        self.value = value
+
+
+class Graph:
+    """An undirected graph of nodes of arbitrary connectivity
+
+    A generic graph class for undirected graphs that holds Nodes and
+    edges that connect them.
+
+    Attributes:
+        nodes(list): A list of Node objects containing the nodes of the graph
+        node_uids(list): A list of unique IDs (uids) for each node
+        uid_to_index(dict): A dictionary that maps UIDs currently present
+            in the graph to their node index
+        neigbors(list of sets):  A list of sets that enumerate the neighbors
+            of each node.  For example the neighbors of node i are the set
+            neighbors[i]
+        next_uid(int):  The next unique ID to be assigned to a node on addition
+    """
+
+    def __init__(self):
+        """Set up an empty graph with no nodes"""
+        self.nodes = []
+        self.node_uids = []
```

Comment:
The UIDs are used internally to make the removal of nodes more efficient.  The ID is the index that they have.